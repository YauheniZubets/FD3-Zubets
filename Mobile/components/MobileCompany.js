import React from 'react';
import PropTypes from 'prop-types';

import MobileClient from './MobileClient';
import AddOrEditClient from './AddClient';
import ModulesClients from '../modules/modules';  // функции с логикой добавления и удаления 

import './MobileCompany.css';

import {addOrEditEvents} from './events';

class MobileCompany extends React.PureComponent {

  constructor (props){
    super(props);
    this.moduleLogic= new ModulesClients; //логика методов: добавления, удаления клиентов
  }

  static propTypes = {
    name: PropTypes.string.isRequired,
    clients:PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.number.isRequired,
        fam: PropTypes.string.isRequired,
        im: PropTypes.string.isRequired,
        otch: PropTypes.string.isRequired,
        balance: PropTypes.number.isRequired,
      })
    ),
  };

  state = {
    name: this.props.name,
    clients: this.props.clients,
    addOrEditComponent: false,
    addOrEditMode: null,
    currentClient: null,
    showList: 0,
  };

  /*
  setName1 = () => {
    this.setState({name:'МТС'});
  };

  setName2 = () => {
    this.setState({name:'Velcom'});
  };
  */
  
  showAddOrEditComponent = () => {
    this.setState({addOrEditComponent: !this.state.addOrEditComponent, addOrEditMode: 1});
  };
  
  addClient = (data) => {
    this.setState({clients: this.moduleLogic.setNewClient(data, this.state.clients), addOrEditComponent: false});
  };

  editClient = (data) => {
    this.setState({clients: this.moduleLogic.editCurrentClient(data, this.state.clients), addOrEditComponent: false, currentClient: null});
  };

  showEditCard = (client) => {
    this.state.addOrEditComponent  
    ? this.setState({addOrEditComponent: false})
    : this.setState({addOrEditComponent: true, addOrEditMode: 2, currentClient: client})
  };

  activeCli = () => {
    this.setState({showList: 2});
  };

  blockedCli = () => {
    this.setState({showList: 1});
  };

  allCli = () => {
    this.setState({showList: 0});
  };

  delCli = (data) => {
    this.setState({clients: this.moduleLogic.deleteCurrentClient(data, this.state.clients)});
  };

  componentDidMount(){
    addOrEditEvents.addListener('AddClient', this.addClient);
    addOrEditEvents.addListener('EditCurrentClient', this.editClient);
    addOrEditEvents.addListener('ShowEditCard', this.showEditCard);
    addOrEditEvents.addListener('DeleteClient', this.delCli);
  };

  componentWillUnmount = () => {
    addOrEditEvents.removeListener('AddClient',this.addClient);
    addOrEditEvents.removeListener('EditCurrentClient',this.editClient);
    addOrEditEvents.removeListener('ShowEditCard', this.showEditCard);
    addOrEditEvents.removeListener('DeleteClient', this.delCli);
  };
  
  render() {

    console.log("MobileCompany render");
    var clientsCode;

    switch (this.state.showList) {
      case 0:
        clientsCode=this.state.clients.map(client=>{
          return <MobileClient key={client.id} client={client} />
        }) 
        break;
      case 1:
        clientsCode=this.state.clients.map(client=>{
          if (client.balance < 0) {
            return <MobileClient key={client.id} client={client} />
          }
        }) 
        break;
        case 2:
          clientsCode=this.state.clients.map(client=>{
            if (client.balance > 0) {
              return <MobileClient key={client.id} client={client} />
            }
          }) 
          break;
      
      default:
        break;
    }

    return (
      <div className='MobileCompany'>
        <div className='MobileCompanyName'>Компания &laquo;{this.state.name}&raquo;</div>
        <div>
          <button onClick={this.allCli} value='all'>Все</button>
          <button onClick={this.activeCli} value='active'>Активные</button>
          <button onClick={this.blockedCli} value='blocked'>Заблокированные</button>
        </div>
        <div className='MobileCompanyClients'>
          <table>
            <tbody>
              <tr>
                <th>Фамилия</th>
                <th>Имя</th>
                <th>Отчество</th>
                <th>Баланс</th>
                <th>Статус</th>
                <th>Редактировать</th>
                <th>Удалить</th>
              </tr>
              {clientsCode}
            </tbody>
          </table>
        </div>
        <input type="button" value="Добавить клиента" onClick={this.showAddOrEditComponent} />
        {
          (this.state.addOrEditComponent) &&
          <AddOrEditClient mode={this.state.addOrEditMode} currentClient={this.state.currentClient}/>
        }
      </div>
    )
    ;
  }
}

export default MobileCompany;
