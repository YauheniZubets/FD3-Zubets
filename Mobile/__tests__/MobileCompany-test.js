'use strict';

import React from 'react';
import renderer from 'react-test-renderer';

import MobileCompany from '../components/MobileCompany';
import AddOrEditClient from '../components/AddClient';

test('рендеринг заблокированных клиентов', () =>{

    const component=renderer.create(
        <MobileCompany 
            name='Velcom'
            clients={[ 
                {id:101, fam:"Иванов", im:"Иван", otch:"Иванович", balance:200}, 
                {id:105, fam:"Сидоров", im:"Сидор", otch:"Сидорович", balance:250}, 
                {id:110, fam:"Петров", im:"Пётр", otch:"Петрович", balance:180},
                {id:120, fam:"Григорьев", im:"Григорий", otch:"Григорьевич", balance:-220},
            ]}
        />
    );

    let componentTree=component.toJSON();
    expect(componentTree).toMatchSnapshot();

    const buttonBlocked = component.root.find( el => el.type=='button' && el.props.value == 'blocked'); 
    buttonBlocked.props.onClick();

    componentTree=component.toJSON();
    expect(componentTree).toMatchSnapshot();
});

test('рендеринг активных клиентов', () =>{

    const component=renderer.create(
        <MobileCompany 
            name='Velcom'
            clients={[ 
                {id:101, fam:"Иванов", im:"Иван", otch:"Иванович", balance:200}, 
                {id:105, fam:"Сидоров", im:"Сидор", otch:"Сидорович", balance:250}, 
                {id:110, fam:"Петров", im:"Пётр", otch:"Петрович", balance:180},
                {id:120, fam:"Григорьев", im:"Григорий", otch:"Григорьевич", balance:-220},
            ]}
        />
    );

    let componentTree=component.toJSON();
    expect(componentTree).toMatchSnapshot();

    const buttonBlocked = component.root.find( el => el.type=='button' && el.props.value == 'active'); 
    buttonBlocked.props.onClick();

    componentTree=component.toJSON();
    expect(componentTree).toMatchSnapshot();
});

