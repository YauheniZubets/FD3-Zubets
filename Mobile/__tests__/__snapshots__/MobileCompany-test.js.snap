// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`рендеринг активных клиентов 1`] = `
<div
  className="MobileCompany"
>
  <div
    className="MobileCompanyName"
  >
    Компания «
    Velcom
    »
  </div>
  <div>
    <button
      onClick={[Function]}
      value="all"
    >
      Все
    </button>
    <button
      onClick={[Function]}
      value="active"
    >
      Активные
    </button>
    <button
      onClick={[Function]}
      value="blocked"
    >
      Заблокированные
    </button>
  </div>
  <div
    className="MobileCompanyClients"
  >
    <table>
      <tbody>
        <tr>
          <th>
            Фамилия
          </th>
          <th>
            Имя
          </th>
          <th>
            Отчество
          </th>
          <th>
            Баланс
          </th>
          <th>
            Статус
          </th>
          <th>
            Редактировать
          </th>
          <th>
            Удалить
          </th>
        </tr>
        <tr
          className="MobileClient"
        >
          <td>
            Иванов
          </td>
          <td>
            Иван
          </td>
          <td>
            Иванович
          </td>
          <td>
            200
          </td>
          <td
            className="MobileClientActive"
          >
            Активный
          </td>
          <td>
            <button
              onClick={[Function]}
              value="Редактировать"
            >
              Редактировать
            </button>
          </td>
          <td>
            <button
              onClick={[Function]}
              value="Удалить"
            >
              Удалить
            </button>
          </td>
        </tr>
        <tr
          className="MobileClient"
        >
          <td>
            Сидоров
          </td>
          <td>
            Сидор
          </td>
          <td>
            Сидорович
          </td>
          <td>
            250
          </td>
          <td
            className="MobileClientActive"
          >
            Активный
          </td>
          <td>
            <button
              onClick={[Function]}
              value="Редактировать"
            >
              Редактировать
            </button>
          </td>
          <td>
            <button
              onClick={[Function]}
              value="Удалить"
            >
              Удалить
            </button>
          </td>
        </tr>
        <tr
          className="MobileClient"
        >
          <td>
            Петров
          </td>
          <td>
            Пётр
          </td>
          <td>
            Петрович
          </td>
          <td>
            180
          </td>
          <td
            className="MobileClientActive"
          >
            Активный
          </td>
          <td>
            <button
              onClick={[Function]}
              value="Редактировать"
            >
              Редактировать
            </button>
          </td>
          <td>
            <button
              onClick={[Function]}
              value="Удалить"
            >
              Удалить
            </button>
          </td>
        </tr>
        <tr
          className="MobileClient"
        >
          <td>
            Григорьев
          </td>
          <td>
            Григорий
          </td>
          <td>
            Григорьевич
          </td>
          <td>
            -220
          </td>
          <td
            className="MobileClientBlocked"
          >
            Заблокирован
          </td>
          <td>
            <button
              onClick={[Function]}
              value="Редактировать"
            >
              Редактировать
            </button>
          </td>
          <td>
            <button
              onClick={[Function]}
              value="Удалить"
            >
              Удалить
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <input
    onClick={[Function]}
    type="button"
    value="Добавить клиента"
  />
</div>
`;

exports[`рендеринг активных клиентов 2`] = `
<div
  className="MobileCompany"
>
  <div
    className="MobileCompanyName"
  >
    Компания «
    Velcom
    »
  </div>
  <div>
    <button
      onClick={[Function]}
      value="all"
    >
      Все
    </button>
    <button
      onClick={[Function]}
      value="active"
    >
      Активные
    </button>
    <button
      onClick={[Function]}
      value="blocked"
    >
      Заблокированные
    </button>
  </div>
  <div
    className="MobileCompanyClients"
  >
    <table>
      <tbody>
        <tr>
          <th>
            Фамилия
          </th>
          <th>
            Имя
          </th>
          <th>
            Отчество
          </th>
          <th>
            Баланс
          </th>
          <th>
            Статус
          </th>
          <th>
            Редактировать
          </th>
          <th>
            Удалить
          </th>
        </tr>
        <tr
          className="MobileClient"
        >
          <td>
            Иванов
          </td>
          <td>
            Иван
          </td>
          <td>
            Иванович
          </td>
          <td>
            200
          </td>
          <td
            className="MobileClientActive"
          >
            Активный
          </td>
          <td>
            <button
              onClick={[Function]}
              value="Редактировать"
            >
              Редактировать
            </button>
          </td>
          <td>
            <button
              onClick={[Function]}
              value="Удалить"
            >
              Удалить
            </button>
          </td>
        </tr>
        <tr
          className="MobileClient"
        >
          <td>
            Сидоров
          </td>
          <td>
            Сидор
          </td>
          <td>
            Сидорович
          </td>
          <td>
            250
          </td>
          <td
            className="MobileClientActive"
          >
            Активный
          </td>
          <td>
            <button
              onClick={[Function]}
              value="Редактировать"
            >
              Редактировать
            </button>
          </td>
          <td>
            <button
              onClick={[Function]}
              value="Удалить"
            >
              Удалить
            </button>
          </td>
        </tr>
        <tr
          className="MobileClient"
        >
          <td>
            Петров
          </td>
          <td>
            Пётр
          </td>
          <td>
            Петрович
          </td>
          <td>
            180
          </td>
          <td
            className="MobileClientActive"
          >
            Активный
          </td>
          <td>
            <button
              onClick={[Function]}
              value="Редактировать"
            >
              Редактировать
            </button>
          </td>
          <td>
            <button
              onClick={[Function]}
              value="Удалить"
            >
              Удалить
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <input
    onClick={[Function]}
    type="button"
    value="Добавить клиента"
  />
</div>
`;

exports[`рендеринг заблокированных клиентов 1`] = `
<div
  className="MobileCompany"
>
  <div
    className="MobileCompanyName"
  >
    Компания «
    Velcom
    »
  </div>
  <div>
    <button
      onClick={[Function]}
      value="all"
    >
      Все
    </button>
    <button
      onClick={[Function]}
      value="active"
    >
      Активные
    </button>
    <button
      onClick={[Function]}
      value="blocked"
    >
      Заблокированные
    </button>
  </div>
  <div
    className="MobileCompanyClients"
  >
    <table>
      <tbody>
        <tr>
          <th>
            Фамилия
          </th>
          <th>
            Имя
          </th>
          <th>
            Отчество
          </th>
          <th>
            Баланс
          </th>
          <th>
            Статус
          </th>
          <th>
            Редактировать
          </th>
          <th>
            Удалить
          </th>
        </tr>
        <tr
          className="MobileClient"
        >
          <td>
            Иванов
          </td>
          <td>
            Иван
          </td>
          <td>
            Иванович
          </td>
          <td>
            200
          </td>
          <td
            className="MobileClientActive"
          >
            Активный
          </td>
          <td>
            <button
              onClick={[Function]}
              value="Редактировать"
            >
              Редактировать
            </button>
          </td>
          <td>
            <button
              onClick={[Function]}
              value="Удалить"
            >
              Удалить
            </button>
          </td>
        </tr>
        <tr
          className="MobileClient"
        >
          <td>
            Сидоров
          </td>
          <td>
            Сидор
          </td>
          <td>
            Сидорович
          </td>
          <td>
            250
          </td>
          <td
            className="MobileClientActive"
          >
            Активный
          </td>
          <td>
            <button
              onClick={[Function]}
              value="Редактировать"
            >
              Редактировать
            </button>
          </td>
          <td>
            <button
              onClick={[Function]}
              value="Удалить"
            >
              Удалить
            </button>
          </td>
        </tr>
        <tr
          className="MobileClient"
        >
          <td>
            Петров
          </td>
          <td>
            Пётр
          </td>
          <td>
            Петрович
          </td>
          <td>
            180
          </td>
          <td
            className="MobileClientActive"
          >
            Активный
          </td>
          <td>
            <button
              onClick={[Function]}
              value="Редактировать"
            >
              Редактировать
            </button>
          </td>
          <td>
            <button
              onClick={[Function]}
              value="Удалить"
            >
              Удалить
            </button>
          </td>
        </tr>
        <tr
          className="MobileClient"
        >
          <td>
            Григорьев
          </td>
          <td>
            Григорий
          </td>
          <td>
            Григорьевич
          </td>
          <td>
            -220
          </td>
          <td
            className="MobileClientBlocked"
          >
            Заблокирован
          </td>
          <td>
            <button
              onClick={[Function]}
              value="Редактировать"
            >
              Редактировать
            </button>
          </td>
          <td>
            <button
              onClick={[Function]}
              value="Удалить"
            >
              Удалить
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <input
    onClick={[Function]}
    type="button"
    value="Добавить клиента"
  />
</div>
`;

exports[`рендеринг заблокированных клиентов 2`] = `
<div
  className="MobileCompany"
>
  <div
    className="MobileCompanyName"
  >
    Компания «
    Velcom
    »
  </div>
  <div>
    <button
      onClick={[Function]}
      value="all"
    >
      Все
    </button>
    <button
      onClick={[Function]}
      value="active"
    >
      Активные
    </button>
    <button
      onClick={[Function]}
      value="blocked"
    >
      Заблокированные
    </button>
  </div>
  <div
    className="MobileCompanyClients"
  >
    <table>
      <tbody>
        <tr>
          <th>
            Фамилия
          </th>
          <th>
            Имя
          </th>
          <th>
            Отчество
          </th>
          <th>
            Баланс
          </th>
          <th>
            Статус
          </th>
          <th>
            Редактировать
          </th>
          <th>
            Удалить
          </th>
        </tr>
        <tr
          className="MobileClient"
        >
          <td>
            Григорьев
          </td>
          <td>
            Григорий
          </td>
          <td>
            Григорьевич
          </td>
          <td>
            -220
          </td>
          <td
            className="MobileClientBlocked"
          >
            Заблокирован
          </td>
          <td>
            <button
              onClick={[Function]}
              value="Редактировать"
            >
              Редактировать
            </button>
          </td>
          <td>
            <button
              onClick={[Function]}
              value="Удалить"
            >
              Удалить
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <input
    onClick={[Function]}
    type="button"
    value="Добавить клиента"
  />
</div>
`;
